compiler/rustc_metadata/src/dependency_format.rs

```diff
@@ -291,10 +291,9 @@ fn attempt_static(tcx: TyCtxt<'_>) -> Option<DependencyList> {
 
     // All crates are available in an rlib format, so we're just going to link
     // everything in explicitly so long as it's actually required.
-    let last_crate = tcx.crates(()).len();
-    let mut ret = (1..last_crate + 1)
-        .map(|cnum| {
-            if tcx.dep_kind(CrateNum::new(cnum)) == CrateDepKind::Explicit {
+    let mut ret = tcx.crates(()).iter()
+        .map(|&cnum| {
```

Comment:
Do we need to skip the leading `CrateNum(0)`?

Reply:
:shrug: `CrateNum(0)` is the local crate so I don't think we need to do any linking for that.

I think that explicitly dealing with the local crate and only iterating over all external ones is probably better than specialcasing the first item of the `crates` iter. But that's a fairly uninformed opinion

Reply:
I don't understand how that code works in the first place? `last_crate` is `tcx.crates(()).len()` and then it creates a `CrateNum(last_crate)`? That should not be a valid `CrateNum`, right?

Reply:
Oh, `tcx.crates(())` only returns _upstream_ `CrateNum`s. That is far from obvious. I think the query should be renamed to `upstream_crate_nums` or something. The same goes for `CStore::crates_untracked`. Would you be up for doing that, @lcnr? (but maybe in a separate PR)

Reply:
`extern_crates` maybe? :thinking: though i prefer doing that in a separate PR, want to merge this as it is blocking some further work

Reply:
I've personally always used the term "upstream" (e.g. see https://github.com/rust-lang/rust/blob/15d9ba0133ce0b35348e1c8367afe00aec841ffa/compiler/rustc_middle/src/query/mod.rs#L1264). "Extern" often refers to something non-Rust. Although `extern crate foo` suggests otherwise `:)`. 

Anyway, I'm fine with doing this in another PR. You can r=me on this one.

Reply:
`upstream` is also fine :laughing: was thinking of `extern crate something` ^^